pkg/build/admission/admission.go

```diff
@@ -0,0 +1,121 @@
+	kclient "github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	"github.com/openshift/origin/pkg/client"
+)
+
+func init() {
+	admission.RegisterPlugin("BuildByStrategy", func(c kclient.Interface, config io.Reader) (admission.Interface, error) {
+		osClient, ok := c.(client.Interface)
+		if !ok {
+			return nil, errors.New("client is not an OpenShift client")
+		}
+		return NewBuildByStrategy(osClient), nil
+	})
+}
+
+type buildByStrategy struct {
+	*admission.Handler
+	client client.Interface
+}
+
+// NewBuildByStrategy returns an admission control for builds that checks
+// on policy based on the build strategy type
+func NewBuildByStrategy(client client.Interface) admission.Interface {
+	return &buildByStrategy{
+		Handler: admission.NewHandler(admission.Create),
+		client:  client,
+	}
+}
+
+func (a *buildByStrategy) Admit(attr admission.Attributes) error {
+	if resource := attr.GetResource(); resource != "builds" && resource != "buildconfigs" {
+		return nil
+	}
+	var err error
+	switch obj := attr.GetObject().(type) {
+	case *buildapi.Build:
+		err = a.checkBuildAuthorization(attr.GetNamespace(), obj, attr.GetUserInfo())
+	case *buildapi.BuildConfig:
+		err = a.checkBuildConfigAuthorization(attr.GetNamespace(), obj, attr.GetUserInfo())
+	}
+
+	return err
+}
+
+func resourceForStrategyType(strategyType buildapi.BuildStrategyType) string {
+	var resource string
+	switch strategyType {
+	case buildapi.DockerBuildStrategyType:
+		resource = authorizationapi.DockerBuildResource
+	case buildapi.CustomBuildStrategyType:
+		resource = authorizationapi.CustomBuildResource
+	case buildapi.SourceBuildStrategyType:
+		resource = authorizationapi.SourceBuildResource
+	}
+	return resource
+
+}
+
+func resourceName(objectMeta kapi.ObjectMeta) string {
+	if len(objectMeta.GenerateName) > 0 {
+		return objectMeta.GenerateName
+	}
+	return objectMeta.Name
+}
+
+func notAllowed(strategyType buildapi.BuildStrategyType) error {
+	return fmt.Errorf("build strategy type %s is not allowed", strategyType)
+}
+
+func (a *buildByStrategy) checkBuildAuthorization(namespace string, build *buildapi.Build, userInfo user.Info) error {
+	subjectAccessReview := &authorizationapi.SubjectAccessReview{
+		Verb:         "create",
+		Resource:     resourceForStrategyType(build.Parameters.Strategy.Type),
+		User:         userInfo.GetName(),
+		Groups:       util.NewStringSet(userInfo.GetGroups()...),
+		Content:      runtime.EmbeddedObject{Object: build},
+		ResourceName: resourceName(build.ObjectMeta),
+	}
+	resp, err := a.client.SubjectAccessReviews(namespace).Create(subjectAccessReview)
+	if err != nil {
+		return err
+	}
+	if !resp.Allowed {
+		build.Status = buildapi.BuildStatusFailed
```

Comment:
So you're allowed to create it, but it is always failed?  Why is that preferable to a forbidden?